infra/filter.rkt

```diff
@@ -4,57 +4,56 @@
-     (subset? (operators-in core) (map string->symbol ops))]
-    [((or 'not-operators 'not-operations) (list ops ...))
-     (define body-ops (operators-in core))
-     (for/and ([op (map string->symbol ops)])
-       (not (set-member? body-ops op)))]
-
-    [('constant (list value))
-     (set-member? (constants-in core) (string->symbol value))]
-    [('constants (list ops ...))
-     (subset? (constants-in core) (map string->symbol ops))]
-    [('not-constants (list ops ...))
-     (define body-ops (constants-in core))
-     (for/and ([op (map string->symbol ops)])
-       (not (set-member? body-ops op)))]
-
-    [((? symbol?) (list))
-     (define prop (string->symbol (format ":~a" type)))
-     (dict-has-key? (property-values core) prop)]
-    [((? symbol?) (list values ...))
-     (define prop (string->symbol (format ":~a" type)))
-     (subset? (set-map (dict-ref (property-values core) prop '()) ~a) values)]
-    [('cites (list values ...))
-     (subset? (map string->symbol values) (append-map set->list (dict-ref (property-values core) ':cite '())))]
-    [(_ _)
-     (raise-user-error 'filter "Unknown filter ~a with ~a arguments" type (length values))]))
-
-(define (filter-body invert? type values stdin-port stdout-port)
-   (define test
-     ((if invert? negate identity)
-      (filter (string->symbol type) values)))
+(define (get-args str)
+  (if (string-contains? str ":")
+             (string-split str ":")
+             '()))
+
+(define/contract ((filter query) core)
+  (-> string? (-> fpcore? boolean?))
+  (match (get-args query)
+    [(list op val)
+      (match (string->symbol op)
+        ['operator
+          (set-member? (operators-in core) (string->symbol val))]
+        [_ (raise-user-error 'filter "Unknown filter operation ~a" op)])]
+    [_ (raise-user-error 'filter "Improperly formatted input string ~a" query)]))
+
+(define ((filter-helper queries) core)
+  (andmap (lambda (q) ((filter q) core)) queries))
+
+(define (filter-body invert? queries in-file out-file stdin-port stdout-port)
+  (define-values (input-port input-port-name)
+    (if (equal? in-file "-")
+        (values stdin-port "stdin")
+        (values (open-input-file in-file #:mode 'text) in-file)))
+  (define test
+    ((if invert? negate identity)
+      (filter-helper queries)))
+  (define output-port
+     (if (equal? out-file "-")
+         stdout-port
+         (open-output-file out-file #:mode 'text #:exists 'truncate)))
    (port-count-lines! (current-input-port))
-   (for ([core (in-port (curry read-fpcore "stdin") (current-input-port))])
+   (for ([core (in-port (curry read-fpcore input-port-name) input-port)])
      (when (test core)
-       (pretty-print core (current-output-port) 1)
-       (newline))))
+       (pretty-print core output-port 1)
+       (newline output-port))))
 
 (module+ main
   (require racket/cmdline)
   (define invert? #f)
+  (define *filter-in* (make-parameter "-"))
+  (define *filter-out* (make-parameter "-"))
```

Comment:
Ahh I see what's weird here. This is a different parameter than the one in `main.rkt`. Should this one be removed?

Reply:
Yeah it's different but this is also a different entry point if people just want to call filter directly so I'm pretty sure it still needs to be there.